File: web/services/owner/building.service.ts
import api from '@/services/api';
import type {
  Building,
  Room,
  CreateBuildingRequest,
  UpdateBuildingRequest,
  CreateRoomRequest,
  UpdateRoomRequest,
} from '@/types/building.types';
import type { ApiResponse, PaginatedResponse } from '@/types/api.types';

export const buildingService = {
  getBuildings: async (): Promise<Building[]> => {
    // TODO: implement
    const response = await api.get('/owner/buildings');
    return response.data;
  },

  getBuildingById: async (id: number): Promise<Building> => {
    // TODO: implement
    const response = await api.get(`/owner/buildings/${id}`);
    return response.data;
  },

  createBuilding: async (data: CreateBuildingRequest): Promise<ApiResponse<Building>> => {
    // TODO: implement
    const response = await api.post('/owner/buildings', data);
    return response.data;
  },

  updateBuilding: async (id: number, data: UpdateBuildingRequest): Promise<ApiResponse<Building>> => {
    // TODO: implement
    const response = await api.put(`/owner/buildings/${id}`, data);
    return response.data;
  },

  deleteBuilding: async (id: number): Promise<ApiResponse<void>> => {
    // TODO: implement
    const response = await api.delete(`/owner/buildings/${id}`);
    return response.data;
  },

  getRooms: async (buildingId: number): Promise<Room[]> => {
    // TODO: implement
    const response = await api.get(`/owner/buildings/${buildingId}/rooms`);
    return response.data;
  },

  createRoom: async (data: CreateRoomRequest): Promise<ApiResponse<Room>> => {
    // TODO: implement
    const response = await api.post('/owner/rooms', data);
    return response.data;
  },

  updateRoom: async (id: number, data: UpdateRoomRequest): Promise<ApiResponse<Room>> => {
    // TODO: implement
    const response = await api.put(`/owner/rooms/${id}`, data);
    return response.data;
  },

  deleteRoom: async (id: number): Promise<ApiResponse<void>> => {
    // TODO: implement
    const response = await api.delete(`/owner/rooms/${id}`);
    return response.data;
  },
};

File: mobile/lib/services/building_service.dart
import '../models/building.dart';
import 'api_client.dart';

class BuildingService {
  final _apiClient = ApiClient();

  // Get all buildings for current owner
  Future<List<Building>> getBuildings() async {
    try {
      final response = await _apiClient.get('/owner/buildings');
      final data = response.data as Map<String, dynamic>;
      final buildings = (data['data'] as List)
          .map((json) => Building.fromJson(json as Map<String, dynamic>))
          .toList();
      return buildings;
    } catch (e) {
      throw Exception('Failed to fetch buildings: $e');
    }
  }

  // Get building by ID
  Future<Building> getBuildingById(int id) async {
    try {
      final response = await _apiClient.get('/owner/buildings/$id');
      final data = response.data as Map<String, dynamic>;
      return Building.fromJson(data['data'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to fetch building: $e');
    }
  }

  // Create new building
  Future<Building> createBuilding({
    required String name,
    required String address,
  }) async {
    try {
      final response = await _apiClient.post(
        '/owner/buildings',
        data: {
          'name': name,
          'address': address,
        },
      );
      final data = response.data as Map<String, dynamic>;
      return Building.fromJson(data['data'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to create building: $e');
    }
  }

  // Update building
  Future<Building> updateBuilding({
    required int id,
    required String name,
    required String address,
  }) async {
    try {
      final response = await _apiClient.put(
        '/owner/buildings/$id',
        data: {
          'name': name,
          'address': address,
        },
      );
      final data = response.data as Map<String, dynamic>;
      return Building.fromJson(data['data'] as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to update building: $e');
    }
  }

  // Delete building
  Future<void> deleteBuilding(int id) async {
    try {
      await _apiClient.delete('/owner/buildings/$id');
    } catch (e) {
      throw Exception('Failed to delete building: $e');
    }
  }
}
